pipelines/shared/python/network.py

```diff
@@ -0,0 +1,181 @@
+            # Save the expanded size (e.g. accounting for UniRef/repnodes) of the cluster
+            if size_by_seq <= 1:
+                singletons[comp_list[0]] = self.idx_label_map[comp_list[0]]
+            else:
+                cluster_size_by_seq[cluster_idx] = size_by_seq
+                cluster_size_by_node[cluster_idx] = size_by_node
+
+        return cluster_size_by_node, cluster_size_by_seq, singletons
+
+    def compute_cluster_numbers(self, cluster_sizes: dict) -> tuple[dict, dict]:
+        """
+        Create cluster numbers based on the sizes in the input dict
+
+        Parameters
+        ----------
+            cluster_sizes
+                mapping of cluster index (from self.components) to size (either number
+                of sequences of number of nodes)
+
+        Returns
+        -------
+            dict mapping node index to cluster number
+            dict mapping original cluster number to computed cluster number
+        """
+        # Order by cluster size
+        cluster_size_order = dict(sorted(cluster_sizes.items(), key=lambda item: item[1], reverse=True))
+        new_cluster_num = 0
+        idx_cluster = {}
+        cluster_num_map = {}  # Map the old numbering (e.g. index into self.components) to the new numbering
+        # Assign numbers; idx is the old cluster (self.components) index, but the dict is sorted based on the value (cluster size)
+        for idx in cluster_size_order:
+            is_singleton = False
+            for node_idx in self.components[idx]:
+                if node_idx not in self.singletons:
+                    idx_cluster[node_idx] = new_cluster_num + 1
+                else:
+                    is_singleton = True
+            if not is_singleton:
+                cluster_num_map[idx] = new_cluster_num + 1
+                new_cluster_num += 1
+        return idx_cluster, cluster_num_map
+
+    def save_cluster_info(self, info_file: str):
+        """
+        Save the cluster size information to a file, formatted as
+        "cluster_num_by_seq\tcluster_size_by_seq\tcluster_num_by_node\tcluster_size_by_node"
+
+        Parameters
+        ----------
+            info_file
+                path to a file to output cluster info to
+        """
+        with open(info_file, "w") as fh:
+            fh.write("cluster_num_by_seq\tcluster_size_by_seq\tcluster_num_by_node\tcluster_size_by_node\n")
+            # The dict idx here represents the old (self.components) cluster number; sorting
+            # in the file will be done by sequence number
+            for idx in self.cluster_num_map_by_seq:
+                cluster_num_seq = self.cluster_num_map_by_seq[idx]
+                cluster_num_node = self.cluster_num_map_by_node[idx]
+                cluster_size_seq = self.cluster_size_by_seq[idx]
+                cluster_size_node = self.cluster_size_by_node[idx]
+                fh.write(f"{cluster_num_seq}\t{cluster_size_seq}\t{cluster_num_node}\t{cluster_size_node}\n")
```

Comment:
I don't have a strong suggestion on this but since everything is going to end up in a TSV, is there a way to construct the data as a table rather than in dictionaries? I might have missed something but for the most part it looks like you aren't doing random insertions/lookups so constructing parallel arrays and inserting into a pandas dataframe or arrow table might retain more of the semantic meaning of the data as its being processed. Just a thought.

Reply:
I'm not exactly happy with the code since it's more of a brute force approach (although I don't consider it a hack).  Because the same clusters are ordered differently depending on metanode+cluster size or just cluster size I am storing the values in separate dictionaries.  I agree it could be different, but I need to move on to other things.  If I have to refactor this code in the future I'll take your suggestions into account.